### PR TITLE
[GH-1] Introducing DateTime library

### PR DESCRIPTION
**Pull request by DanielaIvanova**

### CODE REVIEW

File: libraries/DateTime/contracts/DateTime.aes [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+      origin_year = 1970, 
+      day_in_seconds = 86400,
+      year_in_seconds = 31536000,
+      leap_year_in_seconds = 31622400,
+      hour_in_seconds = 3600,
+      minute_in_seconds = 60
+      }
+
+  public function parse_timestamp(timestamp : int) : date_time = 
+    let dt : date_time = init_date_time(timestamp)
```
Comment (alexeitcekov) [inline libraries/DateTime/contracts/DateTime.aes]: maybe this function is useless.. you set timestamp in 'init_date_time'

File: libraries/DateTime/contracts/DateTime.aes [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    // calculate second by given timestamp
+    let dt : date_time = dt{second = rest(timestamp, 60)}
+
+    // calculate weekday by given timestamp
+    let wd : int = rest(timestamp / state.day_in_seconds + 4, 7)
+    let dt : date_time = dt{weekday = wd}
+    dt
+
+  public function get_year(timestamp : int) : int =
+    parse_timestamp(timestamp).year
```
Comment (alexeitcekov) [inline libraries/DateTime/contracts/DateTime.aes]: maybe you should separate all these calculations. if you get year(get_year) compute only year... if you get try to get month calculate year and month ... and so on

File: libraries/DateTime/test/dateTime.js [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+      const contract = await deployContract(owner)
+      const callWithdraw = await callContract(
+        contract,
+        'parse_timestamp',
+        timestamp
+      )
+      const decodedWithdraw = await callWithdraw
+        .decode('(int, int, int, int, int, int, int, int)')
+        .then(item => item.value.map(el => el.value))
```
Comment (alexeitcekov) [inline libraries/DateTime/test/dateTime.js]: do not use 'async/await' and 'then' together....

Comment (alexeitcekov) [reply libraries/DateTime/test/dateTime.js]: const result = await callWithdraw
        .decode('(int, int, int, int, int, int, int, int)');
        
      const decodedWithdraw =  result.value.map(item => item.value);

File: libraries/DateTime/test/dateTime.js [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+  })
+
+  describe('Test all get functions', async () => {
+    const year = 1988
+    const month = 5
+    const day = 5
+    const hour = 2
+    const minute = 31
+    const second = 04
+    const timestamp = 578802664
```
Comment (alexeitcekov) [inline libraries/DateTime/test/dateTime.js]: use 'real' data not only hardcoded... there can be a test with hardcoded data and others with 'real'
      // const year = 2016
      // const month = 3
      // const day = 7
      // const hour = 8
      // const minute = 58
      // const second = 47
      // const weekday = 1
      // const timestamp = 1457341127

      let date =  new Date(2016, 1, 29, 10, 45, 32); // or new Date() 
      const timestamp = date.getTime();

      const year = date.getUTCFullYear();
      const month = date.getUTCMonth() + 1;
      const day = date.getUTCDate();
      const hour = date.getUTCHours();
      const minute = date.getUTCMinutes();
      const second = date.getUTCSeconds();
      const weekday = date.getDay();

      let aeTimestamp = parseInt(timestamp / 1000); // to AE_timestamp
      const contract = await deployContract(owner)
      const callWithdraw = await callContract(
        contract,
        'parse_timestamp',
        aeTimestamp
      )

File: libraries/DateTime/test/dateTime.js [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+      const callWithdraw = await contract
+        .call('to_timestamp', {
+          args: `(${year}, ${month}, ${day}, ${hour}, ${minute}, ${second})`,
+          options: { ttl: config.ttl },
+          abi: 'sophia',
+        })
+        .then(async item => item.decode('int'))
+      assert.equal(callWithdraw.value, timestamp)
+    })
```
Comment (alexeitcekov) [inline libraries/DateTime/test/dateTime.js]: you may add and this test ...
// let date =  new Date(2016, 1, 29, 10, 45, 32); // or new Date() 
      let date =  new Date() 
      const timestamp = date.getTime();

      const year = date.getUTCFullYear();
      const month = date.getUTCMonth() + 1;
      const day = date.getUTCDate();
      const hour = date.getUTCHours();
      const minute = date.getUTCMinutes();
      const second = date.getUTCSeconds();

      let aeTimestamp = parseInt(timestamp / 1000); // to AE_timestamp
                      
      const callWithdraw = await contract
        .call('to_timestamp', {
          args: `(${year}, ${month}, ${day}, ${hour}, ${minute}, ${second})`,
          options: { ttl: config.ttl },
          abi: 'sophia',
        })
        .then(async item => item.decode('int'))
      assert.equal(callWithdraw.value, aeTimestamp)

File: libraries/DateTime/test/dateTime.js [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+  })
+
+  describe('Test all get functions with multiple data', () => {
+    before(async () => {
+      contract = await deployContract(owner)
+    })
+
+    it('get_year', async () => {
+      const array = [
```
Comment (alexeitcekov) [inline libraries/DateTime/test/dateTime.js]: you can extract this array as constant variable in describe scope..

File: libraries/DateTime/contracts/DateTime.aes (hunk 2 above)
Comment (DanielaIvanova) [reply libraries/DateTime/contracts/DateTime.aes]: Yes, you are right!

File: libraries/DateTime/contracts/DateTime.aes (hunk 1 above)
Comment (DanielaIvanova) [reply libraries/DateTime/contracts/DateTime.aes]: Could you clarify please?

Comment (alexeitcekov) [reply libraries/DateTime/contracts/DateTime.aes]: in init_Date_time you return this record

Comment (DanielaIvanova) [reply libraries/DateTime/contracts/DateTime.aes]: Ok, thank you. `Buff` struct is now removed.

File: libraries/DateTime/test/dateTime.js [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        const second = 7
+        const timestamp = 1152006187
+
+        const callWithdraw = await deployedContract
+          .call('to_timestamp', {
+            args: `(${year}, ${month}, ${day}, ${hour}, ${minute}, ${second})`,
+            options: { ttl: config.ttl },
+            abi: 'sophia'
+          })
+          .then(async item => item.decode('int'))
```
Comment (alexeitcekov) [inline libraries/DateTime/test/dateTime.js]: can you fix 'then'

File: libraries/DateTime/test/dateTime.js [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+
+        timestamp = parseInt(timestamp / 1000)
+
+        const callWithdraw = await deployedContract
+          .call('to_timestamp', {
+            args: `(${year}, ${month}, ${day}, ${hour}, ${minute}, ${second})`,
+            options: { ttl: config.ttl },
+            abi: 'sophia'
+          })
+          .then(async item => item.decode('int'))
```
Comment (alexeitcekov) [inline libraries/DateTime/test/dateTime.js]: can you fix 'then'

**Pull request merged**